### PR TITLE
Dbfs root table migration 334

### LINKED ISSUE #334: Migrate `MANAGED` Parquet/Delta tables from DBFS root to managed location
1. By default we copy the table content (CTAS) or do DEEP CLONE
1. Allow skipping individual tables/databases based on https://github.com/databrickslabs/ucx/pull/680
1. Explore sizing tables or another threshold - see https://github.com/databrickslabs/ucx/issues/330
1. By default we copy the table into a managed table/managed location
1. Allow overriding target to an external table

### PR DESCRIPTION
**Pull request by FastLee**
Closes #334

### CODE REVIEW

**Review by nfx — CHANGES_REQUESTED** [review]
I repeat again: refactoring is out of scope for this PR.

don't mix the incremental functionality of #332 and refactoring in the same PR, otherwise it's impossible to keep track of what's changed and what's added in code review.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 1]
```diff
@@ -55,55 +65,48 @@ def key(self) -> str:
     def kind(self) -> str:
         return "VIEW" if self.view_text is not None else "TABLE"
 
-    def _sql_external(self, catalog):
-        return f"SYNC TABLE {catalog}.{self.database}.{self.name} FROM {self.key};"
+    def sql_external(self, target_key):
+        return f"SYNC TABLE {target_key} FROM {self.key};"
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:69]: I don't think this will support table names with spaces and unicode for names.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 2]
```diff
@@ -55,55 +65,48 @@ def key(self) -> str:
...
-            return self._sql_external(catalog)
-        else:
-            return self._sql_managed(catalog)
-
-    def sql_alter_to(self, catalog):
-        return (
-            f"ALTER {self.kind} {self.key} SET"
-            f" TBLPROPERTIES ('upgraded_to' = '{catalog}.{self.database}.{self.name}');"
-        )
+        return f"CREATE TABLE IF NOT EXISTS {target_key} DEEP CLONE {self.key};"
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:75]: This won't support table names with spaces and unicode for names.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 3]
```diff
@@ -220,21 +215,114 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
+        target_table_key: str,
+    ):
+        super().__init__(ws, backend, source_table, target_table_key)
+
+    def migrate_object(self):
+        source_table = self._source_table
+        target_table_key = self._target_table_key
+        sql = source_table.sql_managed(target_table_key)
+        logger.debug(
+            f"Migrating table {source_table.key} to {target_table_key} (DBFS Root to Managed) using SQL query: {sql}"
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:277]: Everywhere a table name is converted to a string, the table name should be encapsulated in `

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 4]
```diff
@@ -220,21 +215,114 @@ def _describe(self, catalog: str, database: str, table: str) -> Table | None:
...
         self._tc = tc
         self._backend = backend
         self._ws = ws
-        self._database_to_catalog_mapping = database_to_catalog_mapping
-        self._default_catalog = self._init_default_catalog(default_catalog)
         self._seen_tables: dict[str, str] = {}
+        mapping_rules: dict[str, str] = {}
+        for rule in tm.load():
+            mapping_rules[
+                f"hive_metastore.{rule.src_schema}.{rule.src_table}"
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:323]: Encapsulate table names... to handle spaces & unicode.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 5]
```diff
@@ -243,44 +331,58 @@ def _init_default_catalog(default_catalog):
         else:
             return "ucx_default"  # TODO : Fetch current workspace name and append it to the default catalog.
 
+    @staticmethod
+    def sql_alter_to(from_table: Table, to_table_key: str):
+        return f"ALTER {from_table.kind} {from_table.key} SET TBLPROPERTIES ('upgraded_to' = '{to_table_key}');"
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:336]: Encapsulate table names... to handle spaces & unicode.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 6]
```diff
@@ -243,44 +331,58 @@ def _init_default_catalog(default_catalog):
...
-        elif table.object_type == "EXTERNAL":
-            result = next(self._backend.fetch(sql))
-            if result.status_code != "SUCCESS":
-                raise ValueError(result.description)
-            self._backend.execute(table.sql_alter_to(target_catalog))
-            self._backend.execute(table.sql_alter_from(target_catalog))
-            self._seen_tables[target] = table.key
+            target_table = self._mapping_rules.get(table.key)
+            if not target_table:
+                logger.info(f"Table {table.key} doesn't not have a valid mapping and will be skipped")
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:348]: Encapsulate table names... to handle spaces & unicode in table names.

File: src/databricks/labs/ucx/hive_metastore/tables.py [hunk 7]
```diff
@@ -243,44 +331,58 @@ def _init_default_catalog(default_catalog):
...
+    def _migrate_table(self, source_table: Table, target_table: str):
+        if source_table.is_databricks_dataset:
+            msg = (
+                f"Table {source_table.key} is a reference to a databricks "
+                f"dataset {source_table.location} and will not be migrated"
+            )
+            logger.info(msg)
+            return None
+        if self._is_marked_for_skip(source_table):
+            msg = f"Table {source_table.key} is marked to be skipped and will not be upgraded"
```
Comment (dmoore247) [inline src/databricks/labs/ucx/hive_metastore/tables.py:379]: Encapsulate table names... to handle spaces & unicode.

File: tests/unit/hive_metastore/test_migrate.py [hunk 8]
```diff
@@ -17,40 +19,81 @@
 logger = logging.getLogger(__name__)
 
 
-def test_migrate_managed_tables_should_produce_proper_queries():
+def test_migrate_dbfs_root_tables_should_produce_proper_queries():
```
Comment (dmoore247) [inline tests/unit/hive_metastore/test_migrate.py:22]: Please add test for table names, database names with spaces & unicode.

File: tests/unit/hive_metastore/test_migrate.py [hunk 9]
```diff
@@ -17,40 +19,81 @@
...
         "SELECT * FROM hive_metastore.inventory_database.tables",
-        "CREATE TABLE IF NOT EXISTS ucx_default.db1.managed DEEP CLONE hive_metastore.db1.managed;",
-        "ALTER TABLE hive_metastore.db1.managed SET TBLPROPERTIES ('upgraded_to' = 'ucx_default.db1.managed');",
-        "ALTER TABLE ucx_default.db1.managed SET TBLPROPERTIES ('upgraded_from' = 'hive_metastore.db1.managed');",
+        "SHOW TBLPROPERTIES hive_metastore.db1_src.managed_src",
+        "CREATE TABLE IF NOT EXISTS ucx_default.db1_dst.managed_dst DEEP CLONE hive_metastore.db1_src.managed_src;",
+        "ALTER TABLE hive_metastore.db1_src.managed_src SET TBLPROPERTIES "
+        "('upgraded_to' = 'ucx_default.db1_dst.managed_dst');",
+        "ALTER TABLE ucx_default.db1_dst.managed_dst SET TBLPROPERTIES "
+        "('upgraded_from' = 'hive_metastore.db1_src.managed_src');",
```
Comment (dmoore247) [inline tests/unit/hive_metastore/test_migrate.py:87]: Please use `databricks.labs.ucx` namespace for table properties (unless the product has claimed this tag and this is for compatibility.

**Review by dmoore247 — CHANGES_REQUESTED** [review]
Huge effort, thanks for working on this!
My main call out is to make sure to handle table names containing spaces and unicode.

**Comment by nfx** [comment]
let's close this PR in favor of more smaller ones
